[core IR] Add a core decomposition for floor_divide

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110046

## Context

Introduce a core decomposition for `aten.floor_divide` into other `aten` ops, and add it to the core ATen decomposition table.

This replaces the decomposition of `floor_divide` that was used by Inductor. I noticed there was a note on that decomposition

```
# TorchInductor-only decomposition. It should not be taken to core.
# See https://github.com/pytorch/torchdynamo/pull/1120
```

but couldn't discern the reason why this is the case. cc: @lezcano 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

SS-JIA committed:
[core IR] Add a core decomposition for floor_divide

[ghstack-poisoned]

SS-JIA committed:
Update on "[core IR] Add a core decomposition for floor_divide"


## Context

Introduce a core decomposition for `aten.floor_divide` into other `aten` ops, and add it to the core ATen decomposition table.


[ghstack-poisoned]

lezcano commented:
I don't quite remember what was the issue here really. I think that back then, our lowering for `floordiv` in triton was "faster but broken" and people were wary of decomposing `floor_div` as done in core. I think that @peterbell10  worked on fixing the lowering, and now, funnily enough, the decomposition in core is tighter than the lowering in triton (it does not use 2 wheres).

If this was indeed the reason (let's see if Peter remembers) then I guess we can move it to core, yep.

SS-JIA committed:
Update on "[core IR] Add a core decomposition for floor_divide"


## Context

Introduce a core decomposition for `aten.floor_divide` into other `aten` ops, and add it to the core ATen decomposition table.

This replaces the decomposition of `floor_divide` that was used by Inductor. I noticed there was a note on that decomposition

```
# TorchInductor-only decomposition. It should not be taken to core.
# See https://github.com/pytorch/torchdynamo/pull/1120
```

but couldn't discern the reason why this is the case. cc: lezcano 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

SS-JIA commented:
@pytorchbot merge
